filters/auth/grantconfig.go

```diff
@@ -0,0 +1,296 @@
+	// access token.
+	TokenURL string
+
+	// CallbackPath contains the path where the callback requests with the
+	// authorization code should be redirected to.
+	CallbackPath string
+
+	// ClientID, the OAuth2 client id of the current service, used to exchange
+	// the access code. Must be set if ClientIDFile is not provided.
+	ClientID string
+
+	// ClientSecret, the secret associated with the ClientID, used to exchange
+	// the access code. Must be set if ClientSecretFile is not provided.
+	ClientSecret string
+
+	// ClientIDFile, the path to the file containing the OAuth2 client id of
+	// the current service, used to exchange the access code. Must be set if
+	// ClientID is not provided.
+	ClientIDFile string
+
+	// ClientSecretFile, the path to the file containing the secret associated
+	// with the ClientID, used to exchange the access code. Must be set if
+	// ClientSecret is not provided.
+	ClientSecretFile string
+
+	// SecretsProvider is used to read client ID and secret values from the
+	// file system. Supports secret rotation.
+	SecretsProvider secrets.SecretsProvider
+
+	// TokeninfoClient, optional. When set, it will be used for the
+	// authorization requests to TokeninfoURL. When not set, a new default
+	// client is created.
+	TokeninfoClient *authClient
+
+	// AuthClient, optional. When set, it will be used for the
+	// access code exchange requests to TokenURL. When not set, a new default
+	// client is created.
+	AuthClient *net.Client
+
+	// DisableRefresh prevents refreshing the token.
+	DisableRefresh bool
+
+	// AuthURLParameters, optional. Extra URL parameters to add when calling
+	// the OAuth2 authorize or token endpoints.
+	AuthURLParameters map[string]string
+
+	// AccessTokenHeaderName, optional. When set, the access token will be set
+	// on the request to a header with this name.
+	AccessTokenHeaderName string
+
+	// TokeninfoSubjectKey, optional. When set, it is used to look up the subject
+	// ID in the tokeninfo map received from a tokeninfo endpoint request.
+	TokeninfoSubjectKey string
+
+	// TokenCookieName, optional. The name of the cookie used to store the
+	// encrypted access token after a successful token exchange.
+	TokenCookieName string
+
+	// ConnectionTimeout used for tokeninfo endpoint.
+	ConnectionTimeout time.Duration
+
+	// MaxIdleConnectionsPerHost used for tokeninfo endpoint.
+	MaxIdleConnectionsPerHost int
+
+	// Tracer used for tokeninfo endpoint.
+	Tracer opentracing.Tracer
+}
+
+var (
+	ErrMissingClientID        = errors.New("missing client ID")
+	ErrMissingClientSecret    = errors.New("missing client secret")
+	ErrMissingSecretsRegistry = errors.New("missing secrets registry")
+	ErrMissingSecretFile      = errors.New("missing secret file")
+	ErrMissingTokeninfoURL    = errors.New("missing tokeninfo URL")
+	ErrMissingProviderURLs    = errors.New("missing provider URLs")
+)
+
+func (c *OAuthConfig) init() error {
+	if c.initialized {
+		return c.initErr
```

Comment:
`return nil` as initialized = true only if there was not error. Or alternatively on every early error return initialized should be set to true as well.

Overall this feels a bit defensive, I think alternative could be to have public `Init() error` that user of  `OAuthConfig` should call once. Then there is no need to call `err := c.init()` in every `NewGrant*` constructor and 
```go
func initGrant(c *auth.OAuthConfig, o *Options) error {
	grant, err := c.NewGrant()
	if err != nil {
		return err
	}

	grantCallback, err := c.NewGrantCallback()
	if err != nil {
		return err
	}

	grantClaimsQuery, err := c.NewGrantClaimsQuery()
	if err != nil {
		return err
	}

	o.CustomFilters = append(o.CustomFilters, grant, grantCallback, grantClaimsQuery)
	return nil
}
```
could be reduced to
```go
		if err := oauthConfig.Init(); err != nil {
			log.Errorf("Error while initializing oauth grant filters: %v.", err)
		} else {
                        o.CustomFilters = append(o.CustomFilters, oauthConfig.NewGrant(), oauthConfig.NewGrantCallback(), grantClaimsQuery.NewGrantClaimsQuery())
                }
```

Reply:
Fixed. I simplified it as you suggested.